apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt

```diff
@@ -104,6 +137,51 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
     }
   }
 
+  /**
+   * A [BufferedSource] that writes to the given cache sink as it is read.
+   */
+  private class WriteToCacheSource(
+      private val originalBody: Source,
+      private val cacheSink: Sink,
+      private val cacheEditor: DiskLruCache.Editor,
+  ) : Source by originalBody {
+
+    private var hasClosedAndCommitted: Boolean = false
+
+    override fun read(sink: Buffer, byteCount: Long): Long {
+      val buffer = Buffer()
+      val read = try {
+        originalBody.read(buffer, byteCount)
+      } catch (e: Exception) {
+        cacheEditor.abort()
+        throw e
+      }
+
+      if (read == -1L) {
+        // We've read fully, commit the cache edit
+        closeAndCommitCache()
+        return -1L
+      }
+      buffer.peek().readAll(cacheSink)
+      sink.writeAll(buffer)
```

Comment:
Sounds like something that can fail too? Should we abort the editor in that case?